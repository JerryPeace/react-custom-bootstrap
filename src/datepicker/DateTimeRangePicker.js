import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classNames from 'classnames';
import moment from 'moment';
import DateTimePicker from './DateTimePicker';

class DateTimeRangePicker extends Component {
    static propTypes = {
        // fromTime and toTime setting
        fromTimeValue: PropTypes.any,
        fromTimeInputLabel: PropTypes.node,
        toTimeValue: PropTypes.any,
        toTimeInputLabel: PropTypes.node,
        limitRange: PropTypes.number, // unit: sec

        // icons and format
        className: PropTypes.string,
        dateFormat: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]).isRequired,
        fullMonthLabel: PropTypes.bool,
        minTimeGap: PropTypes.number,
        timeFormat: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]).isRequired,
        timeGapUnit: PropTypes.string,

        // methods
        onRangeChange: PropTypes.func,
    };

    static defaultProps = {
        fromTimeValue: moment(),
        fromTimeInputLabel: 'From',
        toTimeValue: moment().add(6, 'days'),
        toTimeInputLabel: 'To',

        dateFormat: 'YYYY-MM-DD',
        fullMonthLabel: true,
        minTimeGap: 1,
        timeFormat: 'HH:mm:ss',
        timeGapUnit: 's',

        onRangeChange: () => {}
    };

    constructor(props) {
        super(props);

        this.state = {
            fromTimeValue: '',
            toTimeValue: ''
        };

        const { dateFormat, timeFormat } = props;
        this.dateTimeFormat = `${dateFormat} ${timeFormat}`;
        this.bindHandlers();
    }

    componentWillReceiveProps(nextProps) {
        const { dateTimeFormat } = this;
        const newFrom = this.getMoment(nextProps.fromTimeValue);
        const newTo = this.getMoment(nextProps.toTimeValue);
        const oldFrom = this.getMoment(this.props.fromTimeValue);
        const oldTo = this.getMoment(this.props.toTimeValue);
        // sync fromTime and toTime state to align props,
        // when only change via props.
        if (!newFrom.isSame(oldFrom) || !newTo.isSame(oldTo)) {
            const {fromTimeValue, toTimeValue} = this.state;
            if (oldFrom.format(dateTimeFormat) === fromTimeValue ||
                oldTo.format(dateTimeFormat) === toTimeValue) {
                this.handleChanged(newFrom, newTo, false);
            }
        }
    }

    bindHandlers() {
        this.getMoment = ::this.getMoment;
        this.getValue = ::this.getValue;
        this.isValidRange = ::this.isValidRange;
        this.isValid = ::this.isValid;
        this.isValidCustomRange = ::this.isValidCustomRange;
        this.handleFromChanged = ::this.handleFromChanged;
        this.handleToChanged = ::this.handleToChanged;
        this.handleChanged = ::this.handleChanged;
        this.isValidToDate = ::this.isValidToDate;
    }

    getMoment(dateTime) {
        return moment(dateTime, this.dateTimeFormat);
    }

    getValue() {
        const { getMoment } = this;
        const stateFromTime = getMoment(this.state.fromTimeValue);
        const stateToTime = getMoment(this.state.toTimeValue);
        const propFromTime = getMoment(this.props.fromTimeValue);
        const propToTime = getMoment(this.props.toTimeValue);

        return {
            from: stateFromTime.isValid() ? stateFromTime : propFromTime,
            to: stateToTime.isValid() ? stateToTime : propToTime
        };
    }

    isValid() {
        const value = this.getValue();
        const isValidFromTime = value.from.isValid();
        const isValidToTime = value.to.isValid();

        return (
            isValidFromTime &&
            isValidToTime &&
            this.isValidRange(value.from, value.to) &&
            this.isValidCustomRange(value.from, value.to)
        );
    }

    isValidRange(momentFromDate, momentToDate) {
        return momentFromDate.isBefore(momentToDate);
    }

    isValidCustomRange(fromTime, toTime) {
        const { limitRange, timeGapUnit } = this.props;
        if (!limitRange) {
            return true;
        }
        return moment(toTime).isBefore(moment(fromTime).add(limitRange, timeGapUnit));
    }

    handleFromChanged(momentFromDate) {
        if (!moment.isMoment(momentFromDate)) {
            return;
        }
        const momentToDate = this.getValue().to;
        this.handleChanged(momentFromDate, momentToDate);
    }

    handleToChanged(momentToDate) {
        if (!moment.isMoment(momentToDate)) {
            return;
        }
        const momentFromDate = this.getValue().from;
        this.handleChanged(momentFromDate, momentToDate);
    }

    handleChanged(momentFromDate, momentToDate, shouldUpdateProps = true) {
        const {
            limitRange,
            minTimeGap,
            timeGapUnit,
            onRangeChange
        } = this.props;
        const { dateTimeFormat, isValidRange, isValidCustomRange } = this;

        const fromTime = momentFromDate.format(dateTimeFormat);
        const minTime = this.getMoment(fromTime).add(minTimeGap, timeGapUnit);

        let toTime = isValidRange(momentFromDate, momentToDate)
            ? momentToDate.format(dateTimeFormat)
            : minTime.format(dateTimeFormat);

        if (limitRange) {
            const maxTime = this.getMoment(fromTime).add(limitRange - 1, timeGapUnit);
            toTime = isValidCustomRange(momentFromDate, momentToDate)
                ? toTime : maxTime.format(dateTimeFormat);
        }

        shouldUpdateProps && onRangeChange(fromTime, toTime);
        this.setState({
            fromTimeValue: fromTime,
            toTimeValue: toTime
        });
    }

    isValidToDate (momentToDate) {
        const momentFromDate = this.getValue().from;
        const isValidRange = momentFromDate.isBefore(moment(momentToDate).add(1, 'd'));

        if (this.props.limitRange) {
            return isValidRange && this.isValidCustomRange(momentFromDate, momentToDate);
        } else {
            return isValidRange;
        }
    }

    render() {
        const {
            className,
            dateFormat,
            fromTimeInputLabel,
            fullMonthLabel,
            minTimeGap,
            timeFormat,
            timeGapUnit,
            toTimeInputLabel,
        } = this.props;
        const { dateTimeFormat, getValue, isValidToDate } = this;

        const fromTime = getValue().from.format(dateTimeFormat);
        const toTime = getValue().to.format(dateTimeFormat);

        return (
            <div className={classNames('rcb-datetime-range-picker', className)}>
                <div className='from'>
                    <DateTimePicker
                        closeOnSelect={false}
                        closeOnTab={false}
                        dateFormat={dateFormat}
                        fullMonthLabel={fullMonthLabel}
                        inputLabel={fromTimeInputLabel}
                        minTimeGap={minTimeGap}
                        open={true}
                        timeFormat={timeFormat}
                        timeGapUnit={timeGapUnit}
                        value={fromTime}
                        onChange={this.handleFromChanged}
                    />
                </div>
                <div className='to'>
                    <DateTimePicker
                        closeOnSelect={false}
                        closeOnTab={false}
                        dateFormat={dateFormat}
                        fullMonthLabel={fullMonthLabel}
                        inputLabel={toTimeInputLabel}
                        isValidDate={isValidToDate}
                        minTimeGap={minTimeGap}
                        open={true}
                        timeFormat={timeFormat}
                        timeGapUnit={timeGapUnit}
                        value={toTime}
                        onChange={this.handleToChanged}
                    />
                </div>
            </div>
        );
    }
}

export default DateTimeRangePicker;
