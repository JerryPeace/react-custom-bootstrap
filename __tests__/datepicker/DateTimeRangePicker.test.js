import React from 'react';
import sinon from 'sinon';
import moment from 'moment';
import {mount, shallow} from 'enzyme';
import DateTimeRangePicker from 'datepicker/DateTimeRangePicker';

describe('DateTimeRangePicker', () => {
    const fromTimeValue = moment('2018-10-11 10:10:10');
    const toTimeValue = moment('2018-10-13 10:10:10');

    test('should render correctly.', () => {
        const dateRangePicker = shallow(
            <DateTimeRangePicker
                fromTimeValue={fromTimeValue}
                toTimeValue={toTimeValue}
            />
        );

        expect(dateRangePicker).toMatchSnapshot();
    });

    test('should be able to get time pair.', () => {
        const dateRangePicker = mount(
            <DateTimeRangePicker
                fromTimeValue={fromTimeValue}
                toTimeValue={toTimeValue}
            />
        );

        const timePairValue = dateRangePicker.node.getValue();
        expect(dateRangePicker.prop('fromTimeValue').isSame(timePairValue.from)).toEqual(true);
        expect(dateRangePicker.prop('toTimeValue').isSame(timePairValue.to)).toEqual(true);
    });

    describe('should be able to change date.', () => {
        let dateRangePicker, onRangeChange;
        const handleDateTimeChangedSpy = sinon.spy(DateTimeRangePicker.prototype, 'handleChanged');

        beforeEach(() => {
            onRangeChange = (fromTime, toTime) => {
                dateRangePicker.setProps({
                    fromTimeValue: moment(fromTime),
                    toTimeValue: moment(toTime)
                });
            };

            dateRangePicker = mount(
                <DateTimeRangePicker
                    fromTimeValue={fromTimeValue}
                    toTimeValue={toTimeValue}
                    onRangeChange={onRangeChange}
                />
            );
        });

        test('changing fromTime calendar.', () => {
            const newfromTimeValue = moment('2018-10-10 20:20:20');

            expect(dateRangePicker.prop('fromTimeValue').isSame(fromTimeValue)).toEqual(true);
            expect(handleDateTimeChangedSpy.callCount).toEqual(0);
            dateRangePicker.node.handleChanged(newfromTimeValue, toTimeValue);
            expect(dateRangePicker.prop('fromTimeValue').isSame(newfromTimeValue)).toEqual(true);
            expect(handleDateTimeChangedSpy.callCount).toEqual(1);
        });

        test('changing toTime calendar.', () => {
            const newToTimeValue = moment('2018-10-20 20:20:20');

            expect(dateRangePicker.prop('toTimeValue').isSame(toTimeValue)).toEqual(true);
            expect(handleDateTimeChangedSpy.callCount).toEqual(1);
            dateRangePicker.node.handleChanged(fromTimeValue, newToTimeValue);
            expect(dateRangePicker.prop('toTimeValue').isSame(newToTimeValue)).toEqual(true);
            expect(handleDateTimeChangedSpy.callCount).toEqual(2);
        });
    });

    describe('should be able to validate time range.', () => {
        const LIMIT_HOURS = 6;

        test('with valid time range.', () => {
            const validTimeRange = {
                from: moment('2018-10-11 10:00:00'),
                to: moment('2018-10-11 14:00:00')
            };
            const dateRangePicker = mount(
                <DateTimeRangePicker
                    fromTimeValue={validTimeRange.from}
                    toTimeValue={validTimeRange.to}
                    limitRange={LIMIT_HOURS * 60 * 60}
                />
            );

            const isValidRange = dateRangePicker.node.isValid();
            expect(isValidRange).toEqual(true);
        });

        test('with invalid time range.', () => {
            const invalidTimeRange = {
                from: moment('2018-10-11 10:00:00'),
                to: moment('2018-10-11 20:00:00')
            };

            const dateRangePicker = mount(
                <DateTimeRangePicker
                    fromTimeValue={invalidTimeRange.from}
                    toTimeValue={invalidTimeRange.to}
                    limitRange={LIMIT_HOURS * 60 * 60}
                />
            );

            const isValidRange = dateRangePicker.node.isValid();
            expect(isValidRange).toEqual(false);
        });
    });
});
