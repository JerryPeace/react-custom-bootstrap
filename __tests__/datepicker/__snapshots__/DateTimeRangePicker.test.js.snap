// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimeRangePicker should render correctly. 1`] = `
<div
  className="rcb-datetime-range-picker"
>
  <div
    className="from"
  >
    <DateTimePicker
      backLabel={
        <TmIcon
          name="calendar"
          prefix="tmicon"
          tag="span"
        />
      }
      closeOnSelect={false}
      closeOnTab={false}
      columnOfMonth={3}
      dateFormat="YYYY-MM-DD"
      disabled={false}
      entryLabel={
        <TmIcon
          name="clock"
          prefix="tmicon"
          tag="span"
        />
      }
      fullMonthLabel={true}
      inputLabel="From"
      inputProps={Object {}}
      isValidDate={[Function]}
      minTimeGap={1}
      onChange={[Function]}
      open={true}
      renderDay={[Function]}
      timeFormat="HH:mm:ss"
      timeGapUnit="s"
      timeToggleLabelDown={
        <TmIcon
          name="chevron-down"
          prefix="tmicon"
          tag="span"
        />
      }
      timeToggleLabelUp={
        <TmIcon
          name="chevron-up"
          prefix="tmicon"
          tag="span"
        />
      }
      value="2018-10-11 10:10:10"
      viewMode="days"
      weekDayDisplayMode="short"
    />
  </div>
  <div
    className="to"
  >
    <DateTimePicker
      backLabel={
        <TmIcon
          name="calendar"
          prefix="tmicon"
          tag="span"
        />
      }
      closeOnSelect={false}
      closeOnTab={false}
      columnOfMonth={3}
      dateFormat="YYYY-MM-DD"
      disabled={false}
      entryLabel={
        <TmIcon
          name="clock"
          prefix="tmicon"
          tag="span"
        />
      }
      fullMonthLabel={true}
      inputLabel="To"
      inputProps={Object {}}
      isValidDate={[Function]}
      minTimeGap={1}
      onChange={[Function]}
      open={true}
      renderDay={[Function]}
      timeFormat="HH:mm:ss"
      timeGapUnit="s"
      timeToggleLabelDown={
        <TmIcon
          name="chevron-down"
          prefix="tmicon"
          tag="span"
        />
      }
      timeToggleLabelUp={
        <TmIcon
          name="chevron-up"
          prefix="tmicon"
          tag="span"
        />
      }
      value="2018-10-13 10:10:10"
      viewMode="days"
      weekDayDisplayMode="short"
    />
  </div>
</div>
`;
